src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReferenceHandling.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Text.Json.Serialization
+{
+    /// <summary>
+    /// This class defines the various ways the <see cref="JsonSerializer"/> can deal with references on serialization and deserialization.
+    /// </summary>
+    public sealed class ReferenceHandling
+    {
+        /// <summary>
+        /// Reference handling semantics will not be applied to complex JSON types when serializing or deserializing.
+        /// </summary>
+        public static ReferenceHandling Default { get; } = new ReferenceHandling(PreserveReferencesHandling.None);
+        /// <summary>
+        /// Reference metadata will be written and honored when deserializing and serializing complex JSON types.
+        /// </summary>
+        public static ReferenceHandling Preserve { get; } = new ReferenceHandling(PreserveReferencesHandling.All);
+
+        private PreserveReferencesHandling _preserveHandlingOnSerialize;
+        private PreserveReferencesHandling _preserveHandlingOnDeserialize;
+
+        /// <summary>
+        /// Creates a new instance of <see cref="ReferenceHandling"/> using the specified <paramref name="handling"/>
+        /// </summary>
+        /// <param name="handling">The specified behavior for write/read preserved references.</param>
+        internal ReferenceHandling(PreserveReferencesHandling handling) : this(handling, handling, ReferenceLoopHandling.Error) { }
+
+        // For future, someone may want to define their own custom Handler with different behaviors of PreserveReferenceHandling on Serialize vs Deserialize and another ReferenceLoopHandling, such as ignore if added in a future.
+        private ReferenceHandling(PreserveReferencesHandling preserveHandlingOnSerialize, PreserveReferencesHandling preserveHandlingOnDeserialize, ReferenceLoopHandling loopHandling)
+        {
+            _preserveHandlingOnSerialize = preserveHandlingOnSerialize;
+            _preserveHandlingOnDeserialize = preserveHandlingOnDeserialize;
+        }
+
+        internal bool ShouldReadPreservedReferences()
+        {
+            return _preserveHandlingOnDeserialize == PreserveReferencesHandling.All;
+        }
+
+        internal bool ShouldWritePreservedReferences()
+        {
+            return _preserveHandlingOnSerialize == PreserveReferencesHandling.All;
+        }
+    }
+
+    internal enum ReferenceLoopHandling
+    {
+        Error = 0,
+    }
+
+
+    /// <summary>
+    /// Defines behaviors to preserve references of JSON complex types.
+    /// </summary>
+    internal enum PreserveReferencesHandling
+    {
+        /// <summary>
+        /// Preserved objects and arrays will not be written/readen.
+        /// </summary>
+        None = 0,
+        /// <summary>
+        /// Preserved objects and arrays will be written/readen.
```

Comment:
What does it mean to be `Preserved`?
```suggestion
        /// Preserved objects and arrays will be written/read.
```

Reply:
To be `Preserved`: Append `$id` and use that as a backpointer (`$ref`) when the same object reference is spotted somewhere else in the object graph.